packages/allow-scripts/src/setup.js

```diff
@@ -186,13 +164,395 @@ function editPackageJson() {
+      )
+
+      binsList.forEach(([name, link]) => {
+        /** @type {BinInfo[]} */
+        const collection = binCandidates.get(name) || []
+        if (collection.length === 0) {
+          binCandidates.set(name, collection)
+        }
+        collection.push({
+          // canonical name for a direct dependency is just dependency name
+          isDirect: directDeps.has(canonicalName),
+          bin: name,
+          path: filePath,
+          link,
+          fullLinkPath: path.relative(rootDir, path.join(filePath, link)),
+          canonicalName,
+        })
+      })
+    }
+  }
+
+  const lavamoatConfig = packageJson.lavamoat || {}
+
+  const configs = {
+    lifecycle: indexLifecycleConfiguration({
+      packagesWithScripts: packagesWithScriptsLifecycle,
+      allowConfig: lavamoatConfig.allowScripts,
+    }),
+    bin: indexBinsConfiguration({
+      binCandidates,
+      allowConfig: lavamoatConfig.allowBins,
+    }),
+  }
+
+  const somePoliciesAreMissing = !!(
+    configs.lifecycle.missingPolicies.length ||
+    configs.bin.somePoliciesAreMissing
+  )
+
+  return {
+    packageJson,
+    configs,
+    somePoliciesAreMissing,
+    canonicalNamesByPath,
+  }
+}
+
+/**
+ * @typedef SetDefaultConfigurationOpts
+ * @property {string} rootDir
+ */
+
+/**
+ * @param {SetDefaultConfigurationOpts} param0
+ */
+async function setDefaultConfiguration({ rootDir }) {
+  const conf = await loadAllPackageConfigurations({ rootDir })
+  const {
+    configs: { lifecycle, bin },
+    somePoliciesAreMissing,
+  } = conf
+
+  console.log('\n@lavamoat/allow-scripts automatically updating configuration')
+
+  if (!somePoliciesAreMissing) {
+    console.log('\nconfiguration looks good as is, no changes necessary')
+    return
+  }
+
+  console.log('\nadding configuration:')
+
+  lifecycle.missingPolicies.forEach((pattern) => {
+    console.log(`- lifecycle ${pattern}`)
+    lifecycle.allowConfig[pattern] = false
+  })
+
+  if (FEATURE.bins && bin.somePoliciesAreMissing) {
+    bin.allowConfig = prepareBinScriptsPolicy(bin.binCandidates)
+    console.log(
+      `- bin scripts linked: ${Object.keys(bin.allowConfig).join(',')}`
+    )
+  }
+
+  // update package json
+  await savePackageConfigurations({
+    rootDir,
+    conf,
+  })
+}
+
+// internals
+
+/**
+ * @param {string} filename
+ * @returns {string}
+ */
+function addInstallParentDir(filename) {
+  const rootDir = process.env.INIT_CWD || process.cwd()
+  return path.join(rootDir, filename)
+}
+
+/**
+ * @param {string} entry
+ * @param {string} file
+ * @returns {boolean}
+ */
+function isEntryPresent(entry, file) {
+  const rcPath = addInstallParentDir(file)
+  if (!existsSync(rcPath)) {
+    return false
+  }
+  const rcFileContents = readFileSync(rcPath, 'utf8')
+  return rcFileContents.includes(entry)
+}
+
 /**
  * @typedef WriteRcFileContentOpts
  * @property {string} file
  * @property {string} entry
  */
 
 /**
- * @typedef AreBinsBlockedOpts
```

Comment:
Note: Previously the inlined typedefs were a bit all over the place. Aside from the `import` of allow-scripts library types in the top, they are now declared just prior to their first usage.